lang/LangPrimSource/PyrFilePrim.cpp

```diff
@@ -195,8 +195,12 @@ int prFileCopy(struct VMGlobals * g, int numArgsPushed)
 
 	const bfs::path& p1 = SC_Codecvt::utf8_str_to_path(filename1);
 	const bfs::path& p2 = SC_Codecvt::utf8_str_to_path(filename2);
-	bfs::copy(p1, p2);
-	return errNone;
+	boost::system::error_code error_code;
+	bfs::copy(p1, p2, error_code);
+	if (error_code)
+		postfl("Warning: %s (copy from \"%s\" to \"%s\")\n", error_code.message().c_str(), filename1, filename2);
+
+	return error_code.value();
```

Comment:
This is also a problem - error_code values don't map onto primitive return values at all - the allowed values are right at the top of the [Writing Primitives](http://doc.sccode.org/Guides/WritingPrimitives.html#Example) guide I linked earlier. The most straightforward thing to do right now (unfortunately) would be to construct the error message, then throw it in an exception. It'll be caught by the interpreter runtime, which will print the message if it's not caught.

The problem with printing like this, even if you do return a nonzero code, is that it's impossible to suppress this message even if the error is caught and handled correctly by the caller. We really need to talk about how we can improve this in the future - inside a primitive, throwing is heavy-handed and unidiomatic, but returning error codes provides very little debugging information for users.

Reply:
what you can see in the primitive opcodes is that you can call an sclang message from where you want to throw the error, e.g. `primitiveFailed`. It's a special symbol in `gSpecialSelectors[primitiveFailed];`